### PR TITLE
Introduce LoadingStrategy for Dod

### PR DESCRIPTION
**Pull request by artyomychsvc**
We need to be able to load the data in some cases from the cache only if it exists(memory or storage) without net requests.

**The exact case**: User opens the screen with some hidden but prefetched(if exists in the storage cache) data. To improve ux and don't make unneeded net api calls. The patch allows us to control net request count on startup.

### CODE REVIEW

**Review by yatsinar — DISMISSED** [review]

File: dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt [hunk 1]
```diff
@@ -98,23 +98,24 @@ class DataObservableDelegate<Params : Any, Domain : Any> constructor(
      * Requests data from network and subscribes to updates
      * (can be triggered by other subscribers or manual cache overrides)
      *
-     * @param forceReload - if true network request will be made even if data exists in caches
+     * @param loadingStrategy - if [LoadingStrategy.ForceReload] - network request will be made even if data exists in caches,
```
Comment (iismagilov) [inline dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt:101]: there are two places with strategy docs - here and in LoadingStrategy class itself. Let's keep it in one place - in LoadingStrategy file

Comment (artyomychsvc) [reply dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt]: done

File: dod/src/main/java/com/revolut/rxdata/dod/LoadingStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package com.revolut.rxdata.dod
+
+/**
+ * [DataObservableDelegate] observation loading strategy
+ */
+sealed class LoadingStrategy(
```
Comment (petersamokhin) [inline dod/src/main/java/com/revolut/rxdata/dod/LoadingStrategy.kt:6]: nitpick: `selaed interface` & `data object`

Comment (artyomychsvc) [reply dod/src/main/java/com/revolut/rxdata/dod/LoadingStrategy.kt:6]: `data object ` requires 1.7.2
`selaed interface` - looks like unneeded here imho

File: dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt [hunk 3]
```diff
@@ -98,23 +98,22 @@ class DataObservableDelegate<Params : Any, Domain : Any> constructor(
...
      */
-    @Suppress("RedundantLambdaArrow")
-    fun observe(params: Params, forceReload: Boolean = true): Observable<Data<Domain>> =
+    fun observe(params: Params, loadingStrategy: LoadingStrategy): Observable<Data<Domain>> =
         Observable.defer {
             val memCache = fromMemory(params)
             val memoryIsEmpty = memCache == null
             val subject = subject(params)
-            val loading = forceReload || memoryIsEmpty || failedNetworkRequests.containsKey(params)
+            val loading = loadingStrategy.refreshMemory || memoryIsEmpty || failedNetworkRequests.containsKey(params)
```
Comment (iismagilov) [inline dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt:108]: shouldn't it be like
`(loadingStrategy.refreshMemory && (memoryIsEmpty && failedNetworkRequests.containsKey(params))`?

Comment (artyomychsvc) [reply dod/src/main/java/com/revolut/rxdata/dod/DataObservableDelegate.kt:108]: I think with that change we can break current behaviour, as at least we should wait for `failedNetworkRequests.containsKey(params)`

**Review by iismagilov — APPROVED** [review]

**Review by yatsinar — APPROVED** [review]

**Review by yatsinar — DISMISSED** [review]

**Review by iismagilov — APPROVED** [review]

**Review by yatsinar — APPROVED** [review]

**Pull request merged**